lib/oxttl/src/lexer.rs

```diff
@@ -823,6 +834,93 @@ impl N3Lexer {
         Ok(Some(c))
     }
 
+    fn recognize_utf16_surrogate_pair(
+        data: &[u8],
+        position: usize,
+    ) -> Result<Option<char>, TokenRecognizerError> {
+        if data.len() < 4 {
+            return Ok(None);
+        }
+
+        let val_high = str_from_utf8(&data[..4], position..position + 6)?;
+        let surrogate_high = u16::from_str_radix(val_high, 16).map_err(|e| {
+            (
+                position..position + 6,
+                format!(
+                    "The escape sequence '\\u{val_high}' is not a valid hexadecimal string: {e}"
+                ),
+            )
+        })?;
+
+        // https://www.unicode.org/glossary/#low_surrogate_code_point
+        if matches!(surrogate_high, 0xDC00..=0xDFFF) {
+            return Err((
+                position..position + 6,
+                format!("UTF-16 low-surrogate escape sequence '\\u{val_high}' must be preceded by a high-surrogate"),
+            )
+                .into());
+        }
+
+        // https://www.unicode.org/glossary/#high_surrogate_code_point
+        if matches!(surrogate_high, 0xD800..=0xDBFF) {
```

Comment:
If false we are returning `Ok(None)` and not an error even if we know it's an error. I believe it will lead to the parser doing an infinite loop on invalid code points.

Reply:
Originally I wanted to fall back to the original error message from `recognize_hex_char` by mapping `Ok(None)` to it. (We cannot do that because `Ok(None)` is actually a signal to the parser)

`recognize_utf16_surrogate_pair` only triggers when `recognize_hex_char` fails. I wonder what error message should `recognize_utf16_surrogate_pair` print when the first escape sequence is not a surrogate character. Currently it is set to print:
```
The escape sequence '\\u{val_high}' is not a UTF-16 surrogate
```
However, I don't think there is another reason why parsing `\uXXXX` can fail though other than it being a surrogate, so maybe this is ok? (It will never trigger anyway)

Reply:
Yes, I also think it's ok. Thank you!